File: 0076-minimum-window-substring/python/Solution.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        minWSize = float("inf")
        minWStart = -1
        minWEnd = -1
        tmap = dict()
        wmap = dict()
        for i in range(len(t)):
            count = 1
            if t[i] in tmap:
                count += tmap[t[i]]
            tmap[t[i]] = count
        n = len(s)
        l = 0
        r = -1
        while l < n and r < n:
            while not self.match(wmap, tmap):
                r += 1
            if r > n - 1:
                break
            c = s[r]
            if c in tmap:
                count = 1
                if c in wmap:
                    count += wmap[c]
                wmap[c] = count
        if self.match(wmap, tmap):
            if minWSize > r - l + 1:
                minWSize = r - l + 1
                minWStart = l
                minWEnd = r
            c = s[l]
            if c in tmap:
                count = wmap[c]
                if count - 1 == 0:
                    wmap.pop(c)
                else:
                    wmap[c] -= 1
            l += 1
        if minWStart == -1:
            return ""
        return s[minWStart:minWEnd+1]
    def match(self, wmap, tmap):
        for key, val in tmap.items():
            if key not in wmap:
                return False
            if val > wmap[key]:
                return False
        return True

File: 0076-minimum-window-substring/cpp/Solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int minWSize = INT_MAX;
        int minWStart = -1;
        int minWEnd = -1;
        unordered_map<char, int> tmap;
        unordered_map<char, int> wmap;
        for (int i = 0; i < t.size(); ++i) {
            int count = 1;
            if (tmap.find(t[i]) != tmap.end()) {
                count += tmap[t[i]];
            }
            tmap[t[i]] = count;
        }

        int n = s.size();
        int l = 0;
        int r = -1;
        while (l < n && r < n) {
            while (!match(wmap, tmap)) {
                r++;
                if (r > n - 1) {
                    break;
                }
                char c = s[r];
                if (tmap.find(c) != tmap.end()) {
                    int count = 1;
                    if (wmap.find(c) != wmap.end()) {
                        count += wmap[c];
                    }
                    wmap[c] = count;
                }
            }
            if (match(wmap, tmap)) {
                if (minWSize > r - l + 1) {
                    minWSize = r - l + 1;
                    minWStart = l;
                    minWEnd = r;
                }
                char c = s[l];
                if (tmap.find(c) != tmap.end()) {
                    int count = wmap[c];
                    if (count - 1 == 0) {
                        wmap.erase(c);
                    } else {
                        wmap[c] = count - 1;
                    }
                }
                l++;
            }
        }
        if (minWStart == -1) return "";
        return s.substr(minWStart, minWEnd + 1 - minWStart);
    }
    bool match(unordered_map<char, int>& wmap, unordered_map<char, int>& tmap) {
        for (auto it = tmap.begin(); it != tmap.end(); ++it) {
            char key = it->first;
            if (wmap.find(key) == wmap.end()) return false;
            if (wmap[key] < it->second) return false;
        }
        return true;
    }
};
